test/util.py

```diff
@@ -6,129 +6,80 @@
-        elif isinstance(y, torch.Tensor) and isinstance(x, Number):
-            self.assertEqual(x, y.item(), prec, message, allow_inf)
-        elif isinstance(x, torch.Tensor) and isinstance(y, numpy.bool_):
-            self.assertEqual(x.item(), y, prec, message, allow_inf)
-        elif isinstance(y, torch.Tensor) and isinstance(x, numpy.bool_):
-            self.assertEqual(x, y.item(), prec, message, allow_inf)
-        elif isinstance(x, torch.Tensor) and isinstance(y, torch.Tensor):
-            self.__assertTensorsEqual(x, y, prec, message, allow_inf)
-        elif isinstance(x, dict) and isinstance(y, dict):
-            if isinstance(x, OrderedDict) and isinstance(y, OrderedDict):
-                self.assertEqual(x.items(), y.items())
-            else:
-                self.assertEqual(set(x.keys()), set(y.keys()))
-                key_list = list(x.keys())
-                self.assertEqual([x[k] for k in key_list], [y[k] for k in key_list])
-        elif is_iterable(x) and is_iterable(y):
-            super(TestCase, self).assertEqual(len(x), len(y), message)
-            for x_, y_ in zip(x, y):
-                self.assertEqual(x_, y_, prec, message)
-        elif isinstance(x, Number) and isinstance(y, Number):
-            if abs(x) == inf or abs(y) == inf:
-                if allow_inf:
-                    super(TestCase, self).assertEqual(x, y, message)
-                else:
-                    self.fail("Expected finite numeric values - x={}, y={}".format(x, y))
-                return
-            super(TestCase, self).assertLessEqual(abs(x - y), prec, message)
-        else:
-            super(TestCase, self).assertEqual(x, y, message)
-
-    def checkTraceTVM(self, func, input_tensors=None, input_shapes=None, size=100000, runs=100, verbose=False):
-        # prepare inputs
-        if input_tensors is None:
-            if input_shapes is None:
-                seed = torch.rand(size) / runs / 2
-                input_tensors = (seed, seed, seed)
-            else:
-                input_tensors = []
-                for shape in input_shapes:
-                    seed = torch.rand(*shape) / runs / 2
-                    input_tensors.append(seed)
+    def __init__(self, method_name="runTest"):
+        super(TVMTest, self).__init__(method_name)
+
+    @classmethod
+    def rand_int(cls, min_=1, max_=4):
+        def get():
+            return random.randint(min_, max_)
+
+        return get
+
+    @classmethod
+    def rand_list(cls, elem_fn, num_elem):
+        def get():
+            return [elem_fn() for _ in range(num_elem)]
+
+        return get
+
+    @classmethod
+    def rand_shape(
+        cls, min_rank=1, max_rank=4, min_dim=1, max_dim=2 ** 4, like=None, rank=None
+    ):
+        def get():
+            rank_ = rank
+            if not rank_:
+                rank_ = cls.rand_int(min_rank, max_rank)()
+            return cls.rand_list(cls.rand_int(min_dim, max_dim), rank_)()
+
+        return get
```

Comment:
shall we also make a rand_tensor? I think it would be beneficial so that we don't need to construct the tensor everyone we wrote the test.

Reply:
follow up diff? :)